icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,462 @@
+// © 2022 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include <utility>
+
+#include "unicode/utypes.h"
+
+#if !UCONFIG_NO_BREAK_ITERATION
+
+#include "cjmlbe.h"
+#include "brkeng.h"
+#include "dictbe.h"
+#include "unicode/uniset.h"
+#include "unicode/chariter.h"
+#include "unicode/resbund.h"
+#include "unicode/ubrk.h"
+#include "unicode/usetiter.h"
+#include "ubrkimpl.h"
+#include "utracimp.h"
+#include "uvectr32.h"
+#include "uvector.h"
+#include "uassert.h"
+#include "unicode/normlzr.h"
+#include "cmemory.h"
+#include "dictionarydata.h"
+
+U_NAMESPACE_BEGIN
+
+#if !UCONFIG_NO_NORMALIZATION
+
+class Element : public UMemory {
+public:
+    /**
+     * Default constructor.
+     *
+     * @param ch A Unicode character
+     * @param index The unicode block index of the character.
+     */
+    Element(UChar32 ch, UnicodeString index) {
+        this->ch = ch;
+        this->index = index;
+    }
+
+    UChar32 ch;
+    UnicodeString index;
+};
+
+CjMLBreakEngine::CjMLBreakEngine(UnicodeSet& fDigitOrOpenPunctuationOrAlphabetSet,
+    UnicodeSet& fClosePunctuationSet, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    this->fDigitOrOpenPunctuationOrAlphabetSet = fDigitOrOpenPunctuationOrAlphabetSet;
+    this->fClosePunctuationSet = fClosePunctuationSet;
+    fUnicodeBlock = new UVector32(status);
+    loadMLparams(status);
+}
+
+CjMLBreakEngine::~CjMLBreakEngine() {
+    fUnicodeBlock->removeAllElements();
+    delete fUnicodeBlock;
+    fModel.removeAll();
+    fDigitOrOpenPunctuationOrAlphabetSet.removeAllStrings();
+    fClosePunctuationSet.removeAllStrings();
+}
+
+static const UChar32 INVALID = '|';
+
+static UnicodeString* concatChar(const char* str, UChar32 arr[], UErrorCode& status) {
+    UnicodeString* temp = new UnicodeString(str);
+    int32_t idx = 0;
+    while (arr[idx] != -1) {
+        temp->append(arr[idx++]);
+    }
+    LocalPointer<UnicodeString> result(temp, status);
+    return result.orphan();
+}
+
+static UnicodeString* concatUnicodeString(const char* str1, UnicodeString &str2, UErrorCode& status) {
+    UnicodeString* temp = new UnicodeString(str1);
+    temp->append(str2);
+    LocalPointer<UnicodeString> result(temp, status);
+    return result.orphan();
+}
+
+int32_t CjMLBreakEngine::divideUpDictionaryRange( UText *inText,
+        int32_t rangeStart,
+        int32_t rangeEnd,
+        UVector32 &foundBreaks,
+        UnicodeString &inString,
+        LocalPointer<UVector32> &inputMap,
+        UErrorCode& status) const {
+
+    if (U_FAILURE(status))
+        return 0;
+    if (rangeStart >= rangeEnd) {
+        return 0;
+    }
+
+    UVector32 boundary(inString.countChar32() + 1, status);
+    int32_t numBreaks = 0;
+    UChar32 ch;
+    UnicodeString index;
+    UVector* elementList = new UVector(6, status);
```

Comment:
Since you need to delete the element in the elementList, instead of deleting by a loop, you should pass in a "UObjectDeleter" pointer and use the  UVector(UObjectDeleter *d, UElementsAreEqual *c, int32_t initialCapacity, UErrorCode &status); constructor instead. and you can just allocate that on the stack as

UVector elementList(&myElementDeleteFunc, &myElementEqualFunc , 6, status);

Where myElementDeleteFunc is a function 
For example, take a look at "strings = new UVector(...)" the function UnicodeSet::allocateStrings()
notice in there, the strings is allocated on the heap but since in your code you always delete elementList in the same function, you can just allocate that on the stack without using "new"

Reply:
You did so in CjMLBreakEngine::evaluateBreakpoint so just follow that, but allocate it on the stack, not by using new since the elementList is local

Reply:
Done. Thanks for the suggestion.